python-runnables/add-nginx-controller/runnable.py

```diff
@@ -0,0 +1,129 @@
+from dataiku.runnables import Runnable
+import dataiku
+import json, logging, os, re, tempfile
+import requests
+from dku_aws.eksctl_command import EksctlCommand
+from dku_aws.aws_command import AwsCommand
+from dku_utils.cluster import get_cluster_from_dss_cluster, get_cluster_generic_property, set_cluster_generic_property
+from dku_utils.access import _has_not_blank_property
+from dku_kube.kubectl_command import run_with_timeout, KubeCommandException
+from dku_utils.access import _has_not_blank_property, _is_none_or_blank
+
+def make_html(command_outputs):
+    divs = []
+    for command_output in command_outputs:
+        cmd_html = '<div>Run: %s</div>' % json.dumps(command_output[0])
+        rv_html = '<div>Returned %s</div>' % command_output[1]
+        out_html = '<div class="alert alert-info"><div>Output</div><pre class="debug" style="max-width: 100%%; max-height: 100%%;">%s</pre></div>' % command_output[2]
+        err_html = '<div class="alert alert-danger"><div>Error</div><pre class="debug" style="max-width: 100%%; max-height: 100%%;">%s</pre></div>' % command_output[3]
+        divs.append(cmd_html)
+        divs.append(rv_html)
+        divs.append(out_html)
+        if command_output[1] != 0 and not _is_none_or_blank(command_output[3]):
+            divs.append(err_html)
+    return '\n'.join(divs)
+
+class InstallNginx(Runnable):
+    """
+    Installs a Nginx ingress controller as described in https://kubernetes.github.io/ingress-nginx/deploy
+    """
+    def __init__(self, project_key, config, plugin_config):
+        self.project_key = project_key
+        self.config = config
+        self.plugin_config = plugin_config
+        
+    def get_progress_target(self):
+        return None
+
+    def run(self, progress_callback):
+        cluster_data, dss_cluster_settings, dss_cluster_config = get_cluster_from_dss_cluster(self.config['clusterId'])
+        
+        if get_cluster_generic_property(dss_cluster_settings, 'nginx-ingress.controller', 'false') == 'true':
+            raise Exception("Nginx controller already installed, remove first")
+
+        kube_config_path = dss_cluster_settings.get_raw()['containerSettings']['executionConfigsGenericOverrides']['kubeConfigPath']
+
+        env = os.environ.copy()
+        env['KUBECONFIG'] = kube_config_path
+        
+        command_outputs = []
+        keep_going = True
+        
+        cmd = ['kubectl', 'apply', '-f', 'https://raw.githubusercontent.com/kubernetes/ingress-nginx/nginx-0.27.1/deploy/static/mandatory.yaml']
```

Comment:
There have been a few [changes in the ingress-nginx repo](https://github.com/kubernetes/ingress-nginx/commit/ab3a789caae65eec4ad6e3b46b19750b481b6bce) that reflect the restructuration of the AWS load balancing offering: the latest version of nginx-ingress now only relies on NLBs (ie L4), the L7 option being replaced by ALBs. Since we already offer the ALB macro, maybe we should refactor the macro a bit ? It would (hopefully) simplify things 🙂